### PR TITLE
OPSEXP-2253 Add healthchecks and conditional service startup based on service health in docker compose

### PR DESCRIPTION
**Pull request by pmacius**
Ref: OPSEXP-2253

### CODE REVIEW

File: docker-compose/compose.yaml [hunk 1]
```diff
@@ -268,9 +339,17 @@ services:
       - "--entryPoints.traefik.address=:8888"
       - "--accesslog=true"
       - "--providers.docker.exposedByDefault=false"
+      - "--ping.entrypoint=web"
```
Comment (gionn) [inline docker-compose/compose.yaml:342]:
```suggestion
```

Screenshot 2024-10-21 at 17 59 04

maybe it's not really needed, looks like just a name for the entrypoint in the config
https://doc.traefik.io/traefik/operations/ping/#entrypoint

File: docker-compose/compose.yaml [hunk 2]
```diff
@@ -268,9 +339,17 @@ services:
...
+      - "--ping=true"
     ports:
       - "8080:8080"
       - "8888:8888"
+    healthcheck:
+      test: ["CMD", "traefik", "healthcheck", "--ping"]
```
Comment (gionn) [inline docker-compose/compose.yaml:348]:
```suggestion
      test: ["CMD", "traefik", "healthcheck"]
```

https://doc.traefik.io/traefik/operations/cli/#healthcheck

doesn't really seems required as per docs

File: docker-compose/compose.yaml [hunk 3]
```diff
@@ -235,6 +294,12 @@ services:
       APP_CONFIG_PLUGIN_PROCESS_SERVICE: "false"
       APP_CONFIG_PLUGIN_MICROSOFT_ONLINE: "false"
       APP_BASE_SHARE_URL: "http://localhost:8080/workspace/#/preview/s"
+    healthcheck:
+      test: ["CMD", "curl", "-f", "http://localhost:8080/"]
+      interval: 10s
+      timeout: 1s
+      retries: 3
+      start_period: 0s
```
Comment (gionn) [inline docker-compose/compose.yaml:302]:
```suggestion
      start_period: 1s
```

File: docker-compose/compose.yaml [hunk 4]
```diff
@@ -213,6 +259,11 @@ services:
       - "5672:5672" # AMQP
       - "61616:61616" # OpenWire
       - "61613:61613" # STOMP
+    healthcheck:
+      test: ["CMD", "/opt/activemq/bin/activemq", "query", "--objname", "type=Broker,brokerName=*,service=Health", "|", "grep", "Good"]
+      interval: 10s
+      timeout: 5s
+      retries: 5
```
Comment (gionn) [inline docker-compose/compose.yaml:266]:
```suggestion
      retries: 5
      start_period: 5s
```

File: docker-compose/compose.yaml [hunk 5]
```diff
@@ -186,6 +226,12 @@ services:
         http://shared-file-store:8099/alfresco/api/-default-/private/sfs/versions/1/file/
       SPRING_ACTIVEMQ_BROKERURL: nio://activemq:61616
       SPRING_ELASTICSEARCH_REST_URIS: http://elasticsearch:9200
+    healthcheck:
+      test: ["CMD", "curl", "-f", "http://localhost:8080/actuator/health"]
+      interval: 20s
+      timeout: 20s
+      retries: 3
+      start_period: 60s
```
Comment (gionn) [inline docker-compose/compose.yaml:234]: does it really take so much time to start with es ready?

File: docker-compose/compose.yaml [hunk 6]
```diff
@@ -170,13 +203,20 @@ services:
     ports:
       - "5601:5601"
     depends_on:
-      - elasticsearch
+      elasticsearch:
+        condition: service_healthy
+    healthcheck:
+      test: ["CMD-SHELL", "curl -s -o /dev/null -w '%{http_code}' http://localhost:5601/status | grep -q '200'"]
```
Comment (gionn) [inline docker-compose/compose.yaml:209]: this should not be different than calling with `curl -f`

File: docker-compose/compose.yaml [hunk 7]
```diff
@@ -145,6 +167,11 @@ services:
     command: postgres -c max_connections=300 -c log_min_messages=LOG
     ports:
       - "5432:5432"
+    healthcheck:
+      test: ["CMD-SHELL", "pg_isready -d $$POSTGRES_DB -U $$POSTGRES_USER"]
+      interval: 10s
+      timeout: 3s
+      retries: 3
```
Comment (gionn) [inline docker-compose/compose.yaml:174]:
```suggestion
      retries: 3
      start_period: 5s
```

File: docker-compose/compose.yaml (hunk 2 above)
Comment (pmacius) [reply docker-compose/compose.yaml:350]: Docs dont mention that but its required
```
# traefik healthcheck
Error calling healthcheck: please enable `ping` to use health check
```

File: docker-compose/compose.yaml (hunk 1 above)
Comment (pmacius) [reply docker-compose/compose.yaml:344]: This file removed or set to traefik doesnt work. Can find any documentation why this happens

File: docker-compose/compose.yaml (hunk 5 above)
Comment (pmacius) [reply docker-compose/compose.yaml]: Changed it to 20s

File: .github/workflows/docker-compose-community.yml [hunk 8]
```diff
@@ -22,6 +22,6 @@ jobs:
     runs-on: ubuntu-latest
     steps:
       - uses: >-
-          Alfresco/alfresco-build-tools/.github/actions/dbp-charts/verify-compose@v7.0.0
+          Alfresco/alfresco-build-tools/.github/actions/dbp-charts/verify-compose@OPSEXP-2255
```
Comment (pmacius) [inline .github/workflows/docker-compose-community.yml:25]:
```suggestion
          Alfresco/alfresco-build-tools/.github/actions/dbp-charts/verify-compose@v8.0.0
```

File: .github/workflows/docker-compose-enterprise.yml [hunk 9]
```diff
@@ -55,7 +55,7 @@ jobs:
     steps:
       - name: Verify docker-compose
         uses: >-
-          Alfresco/alfresco-build-tools/.github/actions/dbp-charts/verify-compose@v7.0.0
+          Alfresco/alfresco-build-tools/.github/actions/dbp-charts/verify-compose@OPSEXP-2255
```
Comment (pmacius) [inline .github/workflows/docker-compose-enterprise.yml:58]:
```suggestion
          Alfresco/alfresco-build-tools/.github/actions/dbp-charts/verify-compose@v8.0.0
```

**Review by gionn — COMMENTED** [review]
LGTM - waiting for the upstream pr merge

**Review by gionn — APPROVED** [review]

File: .github/workflows/docker-compose-community.yml [hunk 10]
```diff
@@ -22,6 +22,6 @@ jobs:
     runs-on: ubuntu-latest
     steps:
       - uses: >-
-          Alfresco/alfresco-build-tools/.github/actions/dbp-charts/verify-compose@v7.0.0
+          Alfresco/alfresco-build-tools/.github/actions/dbp-charts/verify-compose@v8.0.0
```
Comment (gionn) [inline .github/workflows/docker-compose-community.yml:25]: wait for https://github.com/Alfresco/alfresco-build-tools/pull/787

File: docker-compose/compose.yaml [hunk 11]
```diff
@@ -84,8 +80,15 @@ services:
         http://shared-file-store:8099/alfresco/api/-default-/private/sfs/versions/1/file
     ports:
       - "8095:8095"
+    healthcheck:
+      test: ["CMD", "curl", "-f", "http://localhost:8095/actuator/health"]
+      interval: 60s
```
Comment (gionn) [inline docker-compose/compose.yaml:85]: I would avoid more than 30 seconds, it means you risk losing up to 59s for the next check to return healthy

I would try to stick to 10 seconds checks for simpler services and have 30 seconds only for the slowest to start

File: docker-compose/compose.yaml [hunk 12]
```diff
@@ -84,8 +80,15 @@ services:
...
+      timeout: 10s
```
Comment (gionn) [inline docker-compose/compose.yaml:86]: for simple curl it's reasonable to stay low even `2s`

be consistent across different services

File: docker-compose/compose.yaml [hunk 13]
```diff
@@ -98,8 +101,15 @@ services:
...
+    healthcheck:
+      test: ["CMD", "curl", "-f", "http://localhost:8090/ready"]
+      interval: 20s
+      timeout: 10s
+      retries: 3
+      start_period: 10s
     depends_on:
-      - activemq
+      activemq:
+        condition: service_healthy
```
Comment (gionn) [inline docker-compose/compose.yaml:112]: also on filestore?

File: docker-compose/compose.yaml [hunk 14]
```diff
@@ -128,6 +144,12 @@ services:
         -Dalfresco.port=8080
         -Dalfresco.context=alfresco
         -Dalfresco.protocol=http
+    healthcheck:
+      test: ["CMD", "curl", "-f", "http://localhost:8080/share"]
+      interval: 20s
+      timeout: 5s
+      retries: 3
+      start_period: 15s
     labels:
```
Comment (gionn) [inline docker-compose/compose.yaml:153]: for sure it depends on repo

File: docker-compose/compose.yaml [hunk 15]
```diff
@@ -170,22 +204,35 @@ services:
...
     image: quay.io/alfresco/alfresco-elasticsearch-live-indexing:4.1.0
     mem_limit: 1g
     depends_on:
       elasticsearch:
-        condition: service_started
-      search-reindexing:
-        condition: service_completed_successfully
+        condition: service_healthy
+      transform-core-aio:
+        condition: service_healthy
```
Comment (gionn) [inline docker-compose/compose.yaml:222]: also filestore?

I would try to be consistent also with the position of these declaration, let's decide if before or after the healthchecks

File: docker-compose/compose.yaml [hunk 16]
```diff
@@ -222,9 +275,17 @@ services:
...
+      retries: 3
+      start_period: 15s
     depends_on:
-      - activemq
-      - elasticsearch
+      activemq:
+        condition: service_healthy
+      elasticsearch:
+        condition: service_healthy
   digital-workspace:
```
Comment (gionn) [inline docker-compose/compose.yaml:289]: strictly speaking both webapps are dependant on repo - but maybe we should avoid to declare explicit dependencies that when not met will cause unwanted restarts - so ok to not have repo here

**Review by gionn — CHANGES_REQUESTED** [review]

File: .github/workflows/docker-compose-community.yml [hunk 17]
```diff
@@ -22,6 +22,8 @@ jobs:
     runs-on: ubuntu-latest
     steps:
       - uses: >-
-          Alfresco/alfresco-build-tools/.github/actions/dbp-charts/verify-compose@v7.0.0
+          Alfresco/alfresco-build-tools/.github/actions/dbp-charts/verify-compose@v8.1.0
         with:
           compose_file_path: docker-compose/community-compose.yaml
+          postman_path: test/postman/docker-compose
+          postman_json: acs-test-docker-compose-collection.json
```
Comment (pmacius) [inline .github/workflows/docker-compose-community.yml:29]: those are the defaults defined in upstream compose - not sure if to keep them

File: docker-compose/compose.yaml [hunk 18]
```diff
@@ -170,22 +211,37 @@ services:
     ports:
       - "5601:5601"
     depends_on:
-      - elasticsearch
+      elasticsearch:
+        condition: service_healthy
+    healthcheck:
+      test: ["CMD-SHELL", "curl -f http://localhost:5601/status"]
```
Comment (gionn) [inline docker-compose/compose.yaml:217]:
```suggestion
      test: ["CMD", "curl", "-f", "http://localhost:5601/status"]
```

**Review by gionn — APPROVED** [review]

**Pull request merged**